File: its-my-training-day/leetcode-top-75/questions/strings/encoding/encode-decode-strings/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const encodeStrings = (strs: string[]): string => {
  const parts = new ArrayList<string>();
  for (let i = 0; i < strs.length; i += 1) {
    parts.add(String(strs[i].length));
    parts.add("#");
    parts.add(strs[i]);
  }
  return toArray(parts).join("");
};
export const decodeStrings = (s: string): string[] => {
  const result = new ArrayList<string>();
  let i = 0;
  while (i < s.length) {
    let j = i;
    while (s[j] !== "#") j += 1;
    const length = parseInt(s.slice(i, j), 10);
    const start = j + 1;
    result.add(s.slice(start, start + length));
    i = start + length;
  }
  return toArray(result);
};

File: its-my-training-day/leetcode-top-75/questions/strings/encoding/encode-decode-strings/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def encode_strings(strs: List[str]) -> str:
    parts = ArrayList()
    i = 0
    while i < len(strs):
        s = strs[i]
        parts.add(str(len(s)))
        parts.add("#")
        parts.add(s)
        i += 1
    return "".join(parts.to_list())
def decode_strings(s: str) -> List[str]:
    result = ArrayList()
    i = 0
    while i < len(s):
        j = i
        while s[j] != "#":
            j += 1
        length = int(s[i:j])
        start = j + 1
        result.add(s[start:start + length])
        i = start + length
    return result.to_list()
